### ISSUE TITLE
Crashes PHP FPM on my server

### ISSUE THREAD

**Issue by dominic-p** [nextcloud/bookmarks#573]
First, I love the new interface with the images and search (once #568 is handled). Unfortunately, since upgrading to NC 14, whenever I visit the bookmarks app, my server grinds to a halt and I get 502 Bad Gateway errors.

I have a lot of bookmarks (over 1,000) and when I hit the bookmark pages, I see my server logs get _flooded_ with requests (see below).

It seems like some kind of lazy loading system that gets preview images/favicons as you scroll would solve this.

Also, if anyone knows of any PHP FPM config tricks I can use to prevent this type of thing from bringing down the entire pool, I would love to hear them.

### Steps to reproduce
1. Bookmark a lot of pages
2. Visit the bookmarks app page

### Expected behavior
It shouldn't crash PHP FPM

### Actual behavior
It does crash PHP FPM

### Server configuration
**Operating system**: Ubuntu 18.04

**Web server:** nginx

**Database:** MySQL

**PHP version:** 7.0

**Nextcloud version:** 14.0.1

**Bookmarks version:** 0.13.0

**Updated from an older Nextcloud or fresh install:** update

**Signing status:**

```
No errors have been found.
```

**List of activated apps:**

```
Enabled:
  - accessibility: 1.0.1
  - activity: 2.7.0
  - bookmarks: 0.13.0
  - bruteforcesettings: 1.1.0
  - calendar: 1.6.2
  - cloud_federation_api: 0.0.1
  - comments: 1.4.0
  - contacts: 2.1.6
  - dav: 1.6.0
  - federatedfilesharing: 1.4.0
  - federation: 1.4.0
  - files: 1.9.0
  - files_pdfviewer: 1.3.2
  - files_sharing: 1.6.2
  - files_texteditor: 2.6.0
  - files_trashbin: 1.4.1
  - files_versions: 1.7.1
  - files_videoplayer: 1.3.0
  - firstrunwizard: 2.3.0
  - gallery: 18.1.0
  - impersonate: 1.1.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.2.0
  - nextcloud_announcements: 1.3.0
  - notifications: 2.2.1
  - oauth2: 1.2.1
  - password_policy: 1.4.0
  - provisioning_api: 1.4.0
  - serverinfo: 1.4.0
  - sharebymail: 1.4.0
  - support: 1.0.0
  - systemtags: 1.4.0
  - theming: 1.5.0
  - twofactor_backupcodes: 1.3.1
  - updatenotification: 1.4.1
  - workflowengine: 1.4.0
Disabled:
  - admin_audit
  - encryption
  - files_external
  - survey_client
  - user_external
  - user_ldap
```

### Client configuration
**Browser:** Firefox 62

**Operating system:** Windows 10

#### Web server access log
```
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/7506/image HTTP/2.0" 404 3919 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/7507/image HTTP/2.0" 404 3919 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/7507/favicon HTTP/2.0" 200 628 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/8397/image HTTP/2.0" 404 3920 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/8397/favicon HTTP/2.0" 200 1490 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/8398/image HTTP/2.0" 404 3918 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/8399/image HTTP/2.0" 200 109355 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/8398/favicon HTTP/2.0" 200 1231 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/8399/favicon HTTP/2.0" 200 1431 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/8400/favicon HTTP/2.0" 404 3924 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/8400/image HTTP/2.0" 404 3920 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
192.0.2.1 - - [03/Oct/2018:11:27:45 -0700] "GET /apps/bookmarks/bookmark/8401/image HTTP/2.0" 200 9179 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0"
...
a lot more here...
```

In my PHP FPM logs I'll see a line like this

```
[03-Oct-2018 13:25:03] WARNING: [pool nextcloud] server reached pm.max_children setting (10), consider raising it
```

In my PHP FPM slow log I see some traces like the following:

```
[03-Oct-2018 13:27:25]  [pool nextcloud] pid 4086
script_filename = /var/www/nextcloud/index.php
[0x00007f98d72139e0] curl_exec() /var/www/nextcloud/3rdparty/guzzlehttp/guzzle/src/Handler/CurlHandler.php:40
[0x00007f98d7213920] __invoke() /var/www/nextcloud/3rdparty/guzzlehttp/guzzle/src/Handler/Proxy.php:28
[0x00007f98d7213860] GuzzleHttp\Handler\{closure}() /var/www/nextcloud/3rdparty/guzzlehttp/guzzle/src/Handler/Proxy.php:51
[0x00007f98d72137a0] GuzzleHttp\Handler\{closure}() /var/www/nextcloud/3rdparty/guzzlehttp/guzzle/src/PrepareBodyMiddleware.php:37
[0x00007f98d72136b0] __invoke() /var/www/nextcloud/3rdparty/guzzlehttp/guzzle/src/Middleware.php:30
[0x00007f98d72135f0] GuzzleHttp\{closure}() /var/www/nextcloud/3rdparty/guzzlehttp/guzzle/src/RedirectMiddleware.php:70
[0x00007f98d7213540] __invoke() /var/www/nextcloud/3rdparty/guzzlehttp/guzzle/src/RedirectMiddleware.php:107
[0x00007f98d7213460] checkRedirect() /var/www/nextcloud/3rdparty/guzzlehttp/guzzle/src/RedirectMiddleware.php:72
[0x00007f98d72133b0] GuzzleHttp\{closure}() /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/FulfilledPromise.php:39
[0x00007f98d7213250] GuzzleHttp\Promise\{closure}() /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/TaskQueue.php:47
[0x00007f98d72131b0] run() /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php:246
[0x00007f98d7213100] invokeWaitFn() /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php:223
[0x00007f98d7213070] waitIfPending() /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php:267
[0x00007f98d7212fc0] invokeWaitList() /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php:225
[0x00007f98d7212f30] waitIfPending() /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php:62
[0x00007f98d7212e70] wait() /var/www/nextcloud/3rdparty/guzzlehttp/guzzle/src/Client.php:131
[0x00007f98d7212dc0] request() /var/www/nextcloud/lib/private/Http/Client/Client.php:161
[0x00007f98d7212d00] get() /var/www/nextcloud/apps/bookmarks/controller/lib/http/client.php:41
[0x00007f98d7212c20] sendRequest() /var/www/nextcloud/apps/bookmarks/vendor/marcelklehr/link-preview/src/Models/Link.php:60
[0x00007f98d7212b70] fetch() /var/www/nextcloud/apps/bookmarks/vendor/marcelklehr/link-preview/src/Models/Link.php:69
```

#### Nextcloud log (data/nextcloud.log)

Possibly related log entries?
```
{"reqId":"o1muSHcHoHbJPxup6Yya","level":3,"time":"2018-10-03T20:27:05+00:00","remoteAddr":"192.0.2.1","user":"testuser","app":"PHP","method":"GET","url":"\/apps\/bookmarks\/bookmark\/8404\/favicon","message":"Undefined offset: 0 at \/var\/www\/nextcloud\/lib\/private\/Http\/Client\/Response.php#74","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko\/20100101 Firefox\/62.0","version":"14.0.1.1"}

{"reqId":"o1muSHcHoHbJPxup6Yya","level":3,"time":"2018-10-03T20:27:05+00:00","remoteAddr":"192.0.2.1","user":"testuser","app":"index","method":"GET","url":"\/apps\/bookmarks\/bookmark\/8404\/favicon","message":{"Exception":"TypeError","Message":"Return value of OC\\Http\\Client\\Response::getHeader() must be of the type string, null returned","Code":0,"Trace":[{"file":"\/var\/www\/nextcloud\/apps\/bookmarks\/controller\/lib\/previews\/defaultpreviewservice.php","line":143,"function":"getHeader","class":"OC\\Http\\Client\\Response","type":"->","args":["Content-Type"]},{"file":"\/var\/www\/nextcloud\/apps\/bookmarks\/controller\/lib\/previews\/defaultpreviewservice.php","line":113,"function":"fetchImage","class":"OCA\\Bookmarks\\Controller\\Lib\\Previews\\DefaultPreviewService","type":"->","args":["https:\/\/discourse.mailinabox.email\/favicon.ico"]},{"file":"\/var\/www\/nextcloud\/apps\/bookmarks\/controller\/lib\/previews\/faviconpreviewservice.php","line":53,"function":"getOrFetchImageUrl","class":"OCA\\Bookmarks\\Controller\\Lib\\Previews\\DefaultPreviewService","type":"->","args":["https:\/\/discourse.mailinabox.email\/favicon.ico"]},{"file":"\/var\/www\/nextcloud\/apps\/bookmarks\/controller\/rest\/internalbookmarkcontroller.php","line":243,"function":"getImage","class":"OCA\\Bookmarks\\Controller\\Lib\\Previews\\FaviconPreviewService","type":"->","args":[{"id":"8404","url":"https:\/\/discourse.mailinabox.email\/t\/where-do-you-host\/1876","title":"Where do you host? - Mail-in-a-Box Users-Helping-Users","user_id":"testuser","description":"","public":"0","added":"1517012155","lastmodified":"1521717892","clickcount":"0","tags":["Web Resources\/Email"]}]},{"file":"\/var\/www\/nextcloud\/lib\/private\/AppFramework\/Http\/Dispatcher.php","line":166,"function":"getBookmarkFavicon","class":"OCA\\Bookmarks\\Controller\\Rest\\InternalBookmarkController","type":"->","args":[8404]},{"file":"\/var\/www\/nextcloud\/lib\/private\/AppFramework\/Http\/Dispatcher.php","line":99,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"screenshotService":{"apiUrl":"http:\/\/screeenly.com\/api\/v1\/fullsize","__class__":"OCA\\Bookmarks\\Controller\\Lib\\Previews\\ScreenlyPreviewService"},"__class__":"OCA\\Bookmarks\\Controller\\Rest\\InternalBookmarkController"},"getBookmarkFavicon"]},{"file":"\/var\/www\/nextcloud\/lib\/private\/AppFramework\/App.php","line":118,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"screenshotService":{"apiUrl":"http:\/\/screeenly.com\/api\/v1\/fullsize","__class__":"OCA\\Bookmarks\\Controller\\Lib\\Previews\\ScreenlyPreviewService"},"__class__":"OCA\\Bookmarks\\Controller\\Rest\\InternalBookmarkController"},"getBookmarkFavicon"]},{"file":"\/var\/www\/nextcloud\/lib\/private\/AppFramework\/Routing\/RouteActionHandler.php","line":47,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["InternalBookmarkController","getBookmarkFavicon",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"id":"8404","_route":"bookmarks.internal_bookmark.get_bookmark_favicon"}]},{"function":"__invoke","class":"OC\\AppFramework\\Routing\\RouteActionHandler","type":"->","args":[{"id":"8404","_route":"bookmarks.internal_bookmark.get_bookmark_favicon"}]},{"file":"\/var\/www\/nextcloud\/lib\/private\/Route\/Router.php","line":297,"function":"call_user_func","args":[{"__class__":"OC\\AppFramework\\Routing\\RouteActionHandler"},{"id":"8404","_route":"bookmarks.internal_bookmark.get_bookmark_favicon"}]},{"file":"\/var\/www\/nextcloud\/lib\/base.php","line":987,"function":"match","class":"OC\\Route\\Router","type":"->","args":["\/apps\/bookmarks\/bookmark\/8404\/favicon"]},{"file":"\/var\/www\/nextcloud\/index.php","line":42,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"\/var\/www\/nextcloud\/lib\/private\/Http\/Client\/Response.php","Line":74,"CustomMessage":"--"},"userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko\/20100101 Firefox\/62.0","version":"14.0.1.1"}
```

#### Browser log
```
NA
```

**Comment by marcelklehr** [comment]
Uh, that's nasty. Have a look at the `pm.*` directives, specifically `pm.max_children` over here: https://secure.php.net/manual/en/install.fpm.configuration.php

**Comment by dominic-p** [comment]
Thanks for the quick reply. Yeah, I have tried playing around with `pm.max_children`. I'm just afraid that if I set it high enough to handle this kind of load, I'll be putting the entire server at risk.

I think the best way to mitigate it for me for now is going to be with rate limiting at the nginx level. My plan is to configure it so that after first 100 requests or so, the image/favicon requests will start getting 503 errors, but the PHP FPM pool should still be good to go.

I'll give it a try and let you know how it goes.

**Comment by marcelklehr** [comment]
My intuition would have been to decrease max_children, in order to force nginx to wait until a new fpm process becomes available, but I'm not sure if that works as intended.

**Comment by dominic-p** [comment]
Interesting. I've tried a couple values for max_children. In my testing, the lower ones failed faster than the higher ones, but they all failed eventually. I don't think nginx waits for free children. It just shoves requests at the defined upstream until responses get too slow, and then it responds with 504.

I was able to use rate limiting to work around the problem. But now, it will load a chunk of bookmarks and then, when it hits the rate limit, no more bookmarks are loaded and the spinner stays forever. Once search is working again, this shouldn't be an issue. I can also use labels to filter down so the bookmark I need is loaded before the rate limit is hit.

**Comment by marcelklehr** [comment]
It might be a good idea to create a nextcloud background job to fetch those images before they're needed and in small portions.

**Comment by marcelklehr** [comment]
> I haven't looked into the code, but I feel like some kind of lazy loading / infinite scrolling would still be necessary to avoid this type of crash. Just like nextcloud core will load only enough files in a given folder to fill the screen and then load more as you scroll down.

That's actually ~~what's happening~~ what should be happening. When loading the initial bookmarks screen it'll load ~100 bookmarks tops, which each cause a screenshot and a favicon to be fetched from the server. So, 200 requests max. Plus, Browsers have a connection limit of about 10 connections, that means you should only ever get 10 requests at once, iirc. :/

These requests do cause some load, though, as they make the server fetch the images over the interwebs and store them on disk.

**Comment by dominic-p** [comment]
Interesting. Thanks for the clarification. I thought that was how it worked, but that's not what I was seeing in testing. Maybe something is broken in the lazy load code causing it to request images for everything in scope instead of just the first 100?

Also, this is probably only tangentially related, but after an image is stored on disk, does the server still request it from the remote site?

P.S. I think the 10 concurrent request limit might be out dated. Here is a screenshot of my Firefox network waterfall. It scrolls down quite a bit farther than is shown here, and I don't see any delay between the requests (they all seem to be fired at around the same time).

![capture](https://user-images.githubusercontent.com/1195080/46446181-b8a4d280-c72f-11e8-8e90-125ebae91bf6.JPG)

**Comment by marcelklehr** [comment]
You're right. The infinite scroll feature doesn't stop. I also know why, now. Phew.

**Comment by marcelklehr** [comment]
Thanks for your persistence :)

**Comment by dominic-p** [comment]
Thanks for looking into this for me, and so quickly. I really appreciate it!

I'm not 100% sure the fix solves this for me. I manually applied your change from 35005f9 in my instance, and the original problem seems to be fixed (no more loading all bookmark images at once), but after the initial load, scrolling down doesn't fetch any more bookmarks.

I added some console.log() statements to see what was happening, and on initial page load the `infiniteScroll` function gets called maybe 4-7 times. The first time the conditional causes the `this.bookmarks.fetchPage();` to fire. On subsequent calls `this.$el.prop('scrollHeight')` is too high and the function does nothing. So far so good.

The problem comes when I scroll down. The `infiniteScroll` function is never called again. I didn't dig into the code more to determine why, but maybe I should file a separate issue for this?
